### PR TITLE
Second Flashdeposit Test

### PR DESCRIPTION
**Pull request by daryakaviani**
Added a test for 2 consecutive flashdeposits. Found an issue with calculations (potentially using wdiv(1e12) where unnecessary).

### CODE REVIEW

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol [hunk 1]
```diff
@@ -153,6 +157,135 @@ contract FlashBullTestFork is Test {
...
+            ethToCrabInitial,
+            usdcToBorrow,
+            wSqueethToMint
+        );
+
+        vm.startPrank(user1);
+        flashBull.flashDeposit{value: totalEthToBull}(
+            ethToCrabInitial,
+            0,
+            0,
```
Comment (haythem96) [inline packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:198]: I think we can set those to non-zero values, like getting the current prices from the pool and add a bit of slippage

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol [hunk 2]
```diff
@@ -153,6 +157,135 @@ contract FlashBullTestFork is Test {
...
+
+        vm.startPrank(user1);
+        flashBull.flashDeposit{
+            value: calcTotalEthToBull(
+                wethToLendSecond,
+                ethToCrabSecond,
+                usdcToBorrowSecond,
+                wSqueethToMintSecond
+            )
+        }(ethToCrabSecond, 0, 0, 3000);
```
Comment (haythem96) [inline packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:265]: Same here

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol [hunk 3]
```diff
@@ -153,6 +157,135 @@ contract FlashBullTestFork is Test {
...
+                IEulerEToken(eToken)
+                    .balanceOfUnderlying(address(bullStrategy))
+                    .sub(wethToLendFirst)
+            ) <= 1
+        );
+        assertEq(
+            bullStrategy.getCrabBalance().sub(crabToBeMintedSecond),
+            bullCrabBalanceBeforeSecond,
+            "Bull crab balance mismatch for second flashdeposit"
+        );
```
Comment (haythem96) [inline packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:286]: Maybe add assert for user bull balance

**Review by haythem96 — COMMENTED** [review]
LGTM, added few comments.

File: packages/bull-vault/src/LeverageBull.sol [hunk 4]
```diff
@@ -155,8 +155,9 @@ contract LeverageBull {
                 return (ethToLend, usdcToBorrow);
             }
         }
-        return (IEulerEToken(eToken).balanceOfUnderlying(address(this)).wmul(_bullShare).wdiv(ONE.sub(_bullShare)),
-               IEulerDToken(dToken).balanceOf(address(this)).wmul(_bullShare).wdiv(ONE.sub(_bullShare)).div(1e12));
+        uint256 ethToLend = IEulerEToken(eToken).balanceOfUnderlying(address(this)).wmul(_bullShare).wdiv(ONE.sub(_bullShare));
+        uint256 usdcToBorrow = IEulerDToken(dToken).balanceOf(address(this)).wmul(_bullShare).wdiv(ONE.sub(_bullShare));
```
Comment (haythem96) [inline packages/bull-vault/src/LeverageBull.sol:159]: Yeah this make sense, IEulerDToken(dToken).balanceOf() is in 6 decimals already

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol [hunk 5]
```diff
@@ -153,6 +157,135 @@ contract FlashBullTestFork is Test {
...
+        assertEq(
+            IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy)),
+            wethToLend
+        );
+        assertEq(
+            bullStrategy.getCrabBalance().sub(crabToBeMinted),
+            bullCrabBalanceBefore
+        );
+
```
Comment (haythem96) [inline packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:214]: Add an assert for depositor bull balance check

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol [hunk 6]
```diff
@@ -447,60 +630,136 @@ contract FlashBullTestFork is Test {
...
+        ) = _calcCollateralAndBorrowAmount(_crabToDeposit);
 
         IERC20(crabV2).approve(address(bullStrategy), _crabToDeposit);
         bullStrategy.deposit{value: wethToLend}(_crabToDeposit);
 
         return (wethToLend, usdcToBorrow);
     }
 
-    function _calcCollateralAndBorrowAmount(uint256 _crabToDeposit) internal view returns (uint256, uint256) {
+    function _calcCollateralAndBorrowAmount(uint256 _crabToDeposit)
```
Comment (haythem96) [inline packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:648]: The extra 1e12 div is not fixed here

**Review by haythem96 — CHANGES_REQUESTED** [review]
Need to remove the additional 1e12 div in test file

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol (hunk 6 above)
Comment (daryakaviani) [reply packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol]: I fixed it in the else case I believe.

**Review by haythem96 — APPROVED** [review]
LGTM

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol [hunk 7]
```diff
@@ -151,6 +196,197 @@ contract FlashBullTestFork is Test {
             bullStrategy.getCrabBalance().sub(crabToBeMinted),
             bullCrabBalanceBefore
         );
+        assertEq(
```
Comment (alpinechicken) [inline packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:199]: Need to check user eth balance

Comment (alpinechicken) [reply packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:199]: should be getting a small refund

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol [hunk 8]
```diff
@@ -151,6 +196,197 @@ contract FlashBullTestFork is Test {
...
+            "Bull USDC debt amount mismatch for second flashdeposit"
+        );
+        assertTrue(
+            wethToLendSecond.sub(
+                IEulerEToken(eToken)
+                    .balanceOfUnderlying(address(bullStrategy))
+                    .sub(wethToLendFirst)
+            ) <= 1
+        );
+        assertEq(
```
Comment (alpinechicken) [inline packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:385]: Add assert for user eth balance

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol [hunk 9]
```diff
@@ -216,20 +467,31 @@ contract FlashBullTestFork is Test {
...
             "Bull ETH in leverage amount mismatch"
         );
-        assertEq(userBullBalanceBefore.sub(bullToRedeem), bullStrategy.balanceOf(user1), "User1 bull balance mismatch");
         assertEq(
-            crabBalanceBefore.sub(crabToRedeem), crabV2.balanceOf(address(bullStrategy)), "Bull crab balance mismatch"
+            userBullBalanceBefore.sub(bullToRedeem),
+            bullStrategy.balanceOf(user1),
+            "User1 bull balance mismatch"
+        );
+        assertEq(
```
Comment (alpinechicken) [inline packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:475]: We should be able to get an exact value for this if we use uniswap quoter contract get exact eth out for squeeth and usdc. Same for flashdeposit.

File: packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol [hunk 10]
```diff
@@ -279,10 +573,19 @@ contract FlashBullTestFork is Test {
             bullCrabBalanceBefore,
             "Bull crab balance mismatch"
         );
+        assertEq(
+            bullToMint,
+            bullStrategy.balanceOf(user1),
+            "User1 bull balance mismatch"
+        );
```
Comment (alpinechicken) [inline packages/bull-vault/test/integration-test/FlashBullTestFork.t.sol:580]: assert for user eth balance as above

**Review by alpinechicken — CHANGES_REQUESTED** [review]
generally lgtm. I think we probably need an assert for user eth balance for flashWithdraw/flashDeposit because there will generally be refunds

**Pull request merged**